src/firewall.ts

```diff
@@ -460,147 +469,342 @@ class WirefilterFirewallRule implements FirewallRule {
+    addRule(rule: { id: string, expression: string, action: FirewallAction }): this {
+        this.rules.push({
+            id: rule.id,
+            expression: new WirefilterFirewallRule(this.wirefilter, this.scheme, rule.expression),
+            action: rule.action,
+        });
+        return this;
+    }
+
+    /**
+     * Iterates through the list of rules and returns the matching rule id
+     *
+     * @param request incoming request to be matched against
+     * @returns the provided rule id if a successful match is made, undefined otherwise
+     */
+    matchRequest(request: Request): RequestMatchResults {
+        const results = new RulesetMatchResults();
+        const execCtx = ExecutionContext.buildFromRequest(this.wirefilter, this.scheme, this.defaultLists, request);
+        try {
+            for (const rule of this.rules) {
+                if (!rule.expression.matchUsingContext(execCtx)) continue;
+
+                results.matches.push({
+                    ruleId: rule.id,
+                    action: { ...rule.action }
+                });
+
+                if (isTerminalAction(rule.action)) {
+                    break;
+                }
+                if (rule.action.type === 'skip' && rule.action.ruleset === 'current') {
+                    break;
+                }
+            }
+        } finally {
+            execCtx.free();
+        }
+        return results;
+    }
+}
+
+export interface RuleMatch {
+    ruleId: string
+    action: FirewallAction
+}
+
+export interface TerminatingActionMatch {
+    ruleId: string
+    action: TerminalAction
+}
+
+const isTerminatingActionMatch = (match: RuleMatch): match is TerminatingActionMatch => {
+    return isTerminalAction(match.action);
+};
+
+export interface RequestMatchResults {
+    matches: RuleMatch[]
+    skippedPhases: Phase[]
+    skippedProducts: Product[]
+    loggedRules: string[]
+    terminatedEarly: boolean
+    terminalAction: { action: { type: 'no_match' } } | TerminatingActionMatch
+}
+
+class RulesetMatchResults implements RequestMatchResults {
+    matches: RuleMatch[] = [];
+
+    get skippedPhases(): Phase[] {
+        const phases = new Set<Phase>();
+        for (const m of this.matches) {
+            if (m.action.type === 'skip') {
+                m.action.phases?.forEach(p => phases.add(p));
+            }
+        }
+        return Array.from(phases);
+    }
+
+    get skippedProducts(): Product[] {
+        const products = new Set<Product>();
+        for (const m of this.matches) {
+            if (m.action.type === 'skip') {
+                m.action.products?.forEach(p => products.add(p));
+            }
+        }
+        return Array.from(products);
+    }
+
+    get loggedRules(): string[] {
+        return this.matches.filter(m => m.action.type === 'log').map(m => m.ruleId);
+    }
+
+    get terminatedEarly(): boolean {
+        return this.matches.some(m => m.action.type === 'skip' && m.action.ruleset === 'current');
```

Comment:
Couldn't decide on whether matching a terminal action should also set this to true or not, what do you think?

Reply:
I think it should as could lead to confusion if this was false but a log everything rule with expression of `true` wasn't getting matched.

Also whats the reasoning for checking the full list rather than just the last match?

Reply:
Did you mean log everything, or block everything? Logging wouldn't cause the ruleset matching to terminate early right?

Was a bit of a hack to not have to check if the array was empty. Probably could have used `.at` if we're happy to target > node 16.6. Given its over a year EOL, probably safe.

Reply:
did mean log everything and yeah log won't terminate. example was if you had an action sequence of `log -> block -> log` would be weird if `terminatedEarly` was false and the second log action wasn't listed.

Happy to drop support for node 16 given that it doesn't work on node 16 anyway as per https://github.com/SerCeMan/firewalker/issues/21